Mi-go Encampment Improvements and Prisoner Mechanics

Pabblusansky opened the pull request:
#### Summary
Content "Improves mi-go prisoners and adds rescue mission, add new mi-go monsters and mi-go cocoon."

#### Purpose of change
fixes #73217
The existing mi-go prisoner NPCs had limited dialogues and lacked meaningful interaction mechanics (and missions). This PR improves their dialogue system and adds an escort mission for rescued prisoners and a chain of missions to equip them with basic gear (however, I could see that an improvement needs to be made here, maybe the missions are not so deeply constructed), eventually leading them to join the player. 
Also, the NPCs now have the chance to get some traits like "FORGETFUL", "SCHIZOPHRENIA", as they were tortured by the mi-go and their mental state was crippled, and their traits are impacting the first dialogues when they talk with the player.
The main idea here is that now that the prisoners like "pathetic", "feral", and "drained" (as described in the initial request) are actually spawning outside the cages and in the cocoons. Now, in cages, we have the cocoons here which could be opened up by the player and find monsters, prisoners monsters, actual NPCs and other. Added also some mi-go serum which would be randomly applied when the player will try to open the cocoon.

####Description of change
Enhanced dialogue for existing mi-go prisoners with more varied responses
Added condition-specific dialogue based on prisoner mental traits (SCHIZOPHRENIC, FORGETFUL)
Implemented an escort mission where player can bring prisoners to safety
Created snippets for varied dialogue based on prisoner mental state.
Added examine function for the cocoon

When players cut open a mi-go cocoon(he needs a knife to cut the cocoon), they now have a chance to encounter:

Pathetic Prisoners (24% chance): Dazed captives who can communicate but are too disoriented to be rescued(Monster)
Feral Prisoners (6% chance): Seemingly normal prisoners who turn hostile when talked to(Monster)
Drained Captives (15% chance): Nearly lifeless humans who have been irreparably damaged by mi-go experimentation(Monster)
Rescuable NPCs (15% chance): Actual survivors who can be rescued and will join the player(NPC)
Hostile Creatures (15% chance): Monstrous entities that attack immediately upon release
Human Remains (15% chance): Partially dissected corpses, sometimes with belongings
Biomaterials (10% chance): Strange biological components harvested by the mi-go(exisiting items in game)
#### Testing
Tested in-game the changes; the missions are working great, the prisoners are following the player so they would get rescued, and they are interacting with the player with the help of the dialogues. The migo cocoon examine action is working good. 

#### Describe alternatives you've considered
The implementation still needs a proper sprite for the furniture (alien cocoon of the mi-go), and maybe for the civilian monsters("feral","pathetic","drained", and the two new mi-go monsters. Other alternatives are none.
#### Additional context
In a nutshell,These changes are a starting point and not intended to be a perfect solution. I'm open to suggestions for further improvements and ideas.

Pabblusansky committed:
first change mi_go

Pabblusansky committed:
fix spelling error (ups 👼 )

PatrikLundell commented:
Based on the description only (haven't read the changes themselves), I question the viability of having ferals outside of cages, as they'll slaughter everyone else before a PC has time to get inside. Having them locked up would work, but they'd be target practice if the cages are cages rather than cells. High functioning cunning ferals that pretend to be sane in order to be released may work though, although it might be tricky to implement.
Also, mi-go couldn't let ferals out since that would result in them having to fight them, thus killing off the ferals long before the PC has a chance to get there.

Pabblusansky commented:
> Based on the description only (haven't read the changes themselves), I question the viability of having ferals outside of cages, as they'll slaughter everyone else before a PC has time to get inside. Having them locked up would work, but they'd be target practice if the cages are cages rather than cells. High functioning cunning ferals that pretend to be sane in order to be released may work though, although it might be tricky to implement. Also, mi-go couldn't let ferals out since that would result in them having to fight them, thus killing off the ferals long before the PC has a chance to get there.

Initially, the ferals are neutral and are in the mi-go faction; they are not attacked by the mi-go, they become hostile after the dialogue.

Pabblusansky commented:
> I like this. The mi-go prisoners have thus far been fairly basic but now they are a bit more complex and you also sneak in a bit of lore about the mi-go´s. Since you also seem to want some idea´s I can think of a few:
> 
>     1. If the Migo´s are doing experiments on people than how about you show what they might be turning people into? Perhaps something like a human upgraded with Migo tech that acts as a guard ´robot´ or perhaps some grafted together monstrocities that get lose if you open the wrong cage or coccoon. Would be a great bit of enviromental story telling.
> 
>     2. Show the experiments of the Migo on prisoners not just in their mental state but also physically via mutations. Think giving prisoners a high chance of a fairly negative mutation such as frail or genetic instability. Perhaps also give them a chance for a positive mutation such a minor state increase mutation or something like infrared vision. Nothing that changes the morphology of the human too much as that would likely drive them insane or just kill them outright but minor things that shows that the Migo where running some experiments.
> 
>     3. Could we perhaps gain more interaction with the monster prisoners? I think this would be a great opertuinity for players to use the under utilitzed social skill by either convincing them to still follow you and be friendly or to provide you with additional information (Lore) reguarding the migo´s. Perhaps even be able to convert them into some fairly distinct NPC´s types if you manage to keep them safe and feed long enough and pass a couple speech checks. Think the pathetic prisoners having more of those mutations that the migo´s induced in them or the feral once having a bunch of crazy killer traits like cannibal and killer instinct.
> 
>     4. Add some noises that the monster prisoners make, think similar to how the Migo say things in game. You could make both the prisoners and the Migo say similar or some of the same things. This would trick the player as they suddenly aren´t sure if they are hearing a prisoners or migo in the next room and up´s the creep factor.
> 
>     5. perhaps expant the things you can complete the missions with. Does a prisoner really need a jeans specifically or will any long pants do? Do they really need a hunting knife specifically or would a large kitchen knife suffice? And would they really be so insufferable to insist they need a hunting knife if you offered them a combat knife?

Thanks so much for the reply and for the ideas! I'll start trying to implement the main ideas.

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
ADekema implementation of the ideas

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky committed:
Update data/json/monsters/mi-go_prisoners.json

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Pabblusansky commented:
Currently implemented some of [ADekema](https://github.com/ADekema) ideas:

1. Added 2 new mi-go monsters: 
 "mi-go human guard construct" 
    "description": "A horrifying amalgamation of human and mi-go technology.  This once-human subject has been heavily modified with crystalline implants and biomechanical components.  It moves with mechanical precision and seems to be following some kind of programmed directive.",
    "name": { "str": "mi-go experiment" },
    "description": "A grotesque fusion of human and alien flesh.  This unfortunate subject appears to have been modified extensively with grafted tissues and organs.  Multiple limbs protrude at odd angles, and parts of its skin have been replaced with chitinous plating. It moves erratically and emits distorted sounds that might once have been human speech.",

2. added sounds for the mi-go human guard construct, so it would say it sometimes as mi-go, but more cyborg styled. same for mi-go experiment. Changed that some of monster prisoners also would duplicate the mi-go replics and would emit same sounds. 

    
3. Added also some minor mi-go related mutations for the npc, and some dialogues related to that.

Currently I would try to do some examining function for the cocoon, to implement better logic and a more desired gameplay in the c++ structure of the game, and if it would not work - just sit with the changes we have now. So I am changing this PR to the draft at the moment. 
Also, I am really thinking about doing more interactions with the monster prisoners, as we would maybe just  count them as extras. There is no problem to add more dialogues, the main problem is changing the prisoner, who is a monster, at the end of the interaction chain in the NPC, because:
1. It is difficult, as I saw in the docs, to make a monster follow you, be friendly with you, and etc.
2. Change the monster prisoner in an NPC, that might not include a lot of in-game conditions. In a nutshell, this is very difficult and unworthy.
    
Also I tried to change the missions so that they would accept every item type of the backpack, knife, pants and etc. However, this doesn't work, because  mgoal_find_item_group is obsolete, the second idea how to do it doesn't work properly ( or i am too dumb to implement it correctly ) so I would just leave it as it is, maybe try something in future to do with this.

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
Added iexamine.cpp function for the mi-go cocoon.

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky commented:
I've implemented a complete overhaul of the mi-go cocoon, adding an examining function to create a more diverse and immersive gameplay experience.

The cocoon examination function has been written in C++ to provide a more robust interaction system. When players cut open a mi-go cocoon(he needs a knife to cut the cocoon), they now have a chance to encounter:

Pathetic Prisoners (24% chance): Dazed captives who can communicate but are too disoriented to be rescued(Monster)
Feral Prisoners (6% chance): Seemingly normal prisoners who turn hostile when released(Monster)
Drained Captives (15% chance): Nearly lifeless humans who have been irreparably damaged by mi-go experimentation(Monster)
Rescuable NPCs (15% chance): Actual survivors who can be rescued and will join the player(NPC)
Hostile Creatures (15% chance): Monstrous entities that attack immediately upon release(Including new mi-go monsters)
Human Remains (15% chance): Partially dissected corpses, sometimes with belongings
Biomaterials (10% chance): Strange biological components harvested by the mi-go

So now I'm completly rewritting everything as I was not confident what I would make an examine action.. Now the prisoners would spawn in cocoons and outside of the cages too ( I am doubtely about it but I think it's a good point ).

Pabblusansky committed:
fast spelling fix

Pabblusansky committed:
Update src/iexamine.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

PatrikLundell commented:
This new cocoon logic ought to have the side effect of you not being able to see that prisoners are there from the overmap (because they don't exist in the game until they're revealed), which I think is a good thing.

A potential issue is a quest where a character is to be rescued from a Mi-Go site. It would be odd if everyone else is bundled up except this character, so the quest would probably need the captive stuck in a cocoon either with a 100% overriding probability, or placed in a special cocoon. This point may not apply if you still have prisoner locked up conventionally, though.

Pabblusansky committed:
astyle fixing

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky commented:
> A potential issue is a quest where a character is to be rescued from a Mi-Go site. It would be odd if everyone else is bundled up except this character, so the quest would probably need the captive stuck in a cocoon either with a 100% overriding probability, or placed in a special cocoon. This point may not apply if you still have prisoner locked up conventionally, though.

You mean there's an existing mission where you have to rescue a prisoner, and our changes will break the logic of the mission?

PatrikLundell commented:
Yes. It's an Isherwood mission, and I think the captive is called Barry. I've never completed that mission myself, so I'm a bit hazy about the details.

Pabblusansky committed:
lint fixes

Pabblusansky commented:
> Yes. It's an Isherwood mission, and I think the captive is called Barry. I've never completed that mission myself, so I'm a bit hazy about the details.

I made a quick research and found that Barry already has his own unique mapgen implementation in the mi-go tower, and his mission operates independently from our new cocoon system.

I'll leave the Barry mission untouched since they don't intersect, and our changes will only affect random prisoners in cocoons.

Pabblusansky committed:
fina lint fix i guess

Pabblusansky committed:
last last fix i think

Pabblusansky committed:
change from ... to ellipsis

Pabblusansky committed:
more changes for the text

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

ADekema commented:
I have looked at the implementation of the new monsters and like them very much but I think there is a problem lore problem with the havest lists and such from the two new monsters (mi-go human guard construct and migo experiment).

 Basically they shouldn´t have tainted flesh as havestable as those are specific to zombie and not other types of creatures (Unless they are actually zombies that the mi-go are experimenting with). They should probably just drop human parts and probably some migo parts as migo tech is completely organic and a survivor wouldn´t be able to disurn what any of it is any more than that of what the body parts of a normal migo are. Maybe have some or all of the organs besides the brain be missing to show extreme migo experimentation.

The other problem is that the mi-go human guard construct somehow drops metal and has harvestable cybernetics. This goes completely agianst thier technology being organic. If a survivor is to disect one of these they would probably find that it is clearly a human body but large parts of it have been altered or just been replaced entirely with migo technology. And the technology itself is from the survivors perspective made from the same organic materials as normal migo themselves and that is all he can discurn. So I would replace the metal and cybernetics with migo butchery parts on the harvest list.

Pabblusansky commented:
> I have looked at the implementation of the new monsters and like them very much but I think there is a problem lore problem with the havest lists and such from the two new monsters (mi-go human guard construct and migo experiment).
> 
> Basically they shouldn´t have tainted flesh as havestable as those are specific to zombie and not other types of creatures (Unless they are actually zombies that the mi-go are experimenting with). They should probably just drop human parts and probably some migo butchery parts for the cybernetics as migo tech is completely organic and a survivor wouldn´t be able to meaningfully tell the differance between a migo cybernetic put onto a human and a body part of a migo itself. Maybe have some or all of the organs besides the brain be missing to show extreme migo experimentation and replacement of organs with migo tech.
> 
> The other problem is that the mi-go human guard construct somehow drops metal and has harvestable cybernetics. This goes completely agianst thier technology being organic. If a survivor is to disect one of these they would probably find that it is clearly a human body but large parts of it have been altered or just been replaced entirely with migo cybernetics. And the technology itself is from the survivors perspective made from the same organic materials as normal migo themselves as their tech and biomatter are one and the same. So I would replace the metal and cybernetics with migo butchery parts on the harvest list. Maybe also add a few messages about what the survivor finds as you take it apart similar to how gaint insects have extra messages about what the survivor finds when they are butchered.

Thank you a lot, that's a good point. I'm looking to fix this

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
fixing monsters

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky committed:
stop mi-go construct to be cyborg

Pabblusansky commented:
Fixed these two new monsters, mi-go guard construct now it's just a mutant, fixed items on butchering.

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
lint (again)

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky committed:
fix the iexamine.cpp

Pabblusansky committed:
fix iexamine.cpp

Pabblusansky committed:
fix the lint(last time i guess)

Pabblusansky committed:
fix the iexamine.cpp again

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
fix lint!!!

Pabblusansky committed:
AGAIN IEXAMINE.CPP

Pabblusansky committed:
include monster.h

Pabblusansky committed:
fix iexamine again

I'm tired boss...

Pabblusansky committed:
Update src/iexamine.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Pabblusansky committed:
FIX IEXAMINE

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Pabblusansky committed:
lint fix

Pabblusansky committed:
Update iexamine.cpp

Pabblusansky committed:
Update iexamine.cpp

Pabblusansky committed:
fix iexamine

Pabblusansky committed:
fix again

Pabblusansky committed:
fix color to pass the tests

Maleclypse commented on data/json/furniture_and_terrain/furniture-alien.json:
```diff
@@ -816,5 +816,63 @@
     }
+  },
+  {
+    "type": "furniture",
```
```suggestion
    "type": "furniture"
```

Maleclypse committed:
Update data/json/furniture_and_terrain/furniture-alien.json

Maleclypse commented on data/json/furniture_and_terrain/furniture-alien.json:
```diff
@@ -816,5 +816,63 @@
     }
+  },
+  {
+    "type": "furniture"
```
```suggestion
    "type": "furniture",
```

Maleclypse committed:
Update data/json/furniture_and_terrain/furniture-alien.json

Maleclypse commented on data/json/monsters/mi-go_prisoners.json:
```diff
@@ -0,0 +1,75 @@
+    "flags": [ "SEES", "HEARS", "SMELLS", "WARM", "BASHES", "GROUP_MORALE", "HUMAN", "GUILT_HUMAN", "CONVERSATION" ]
+  },
+  {
+    "id": "mon_feral_prisoner",
```
```suggestion
    "id": "mon_feral_abductee",
```

Maleclypse committed:
Update data/json/monsters/mi-go_prisoners.json

Maleclypse commented:
Damnit, you've used Chatgpt to write this.

Maleclypse commented:
I was going to let you try to fix this but I've realized now that you never even read this before submitting it.

Maleclypse commented:
@Pabblusansky if you didn't use chatgpt to write this I'm going to need an explanation of the mutations changes and how you've tested them.

Pabblusansky committed:
Merge branch 'CleverRaven:master' into mi_go_changes

Pabblusansky committed:
Cleaning everything

Pabblusansky committed:
Merge branch 'mi_go_changes' of https://github.com/Pabblusansky/Cataclysm-DDA into mi_go_changes

Maleclypse commented:
Thanks for chatting with us about the LLM policy and good luck on this!

Pabblusansky commented:
The changes were made: deleted the new mi-go monsters, speech for them, and deleted mi-go mutations as not lore-friendly. The trick question is: does the vanilla mutation need to remain, and the dialogues related to this too?

Maleclypse commented on data/json/speech.json:
```diff
@@ -49,31 +49,31 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go", "mon_mi_go_slaver" ],
+    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_pathetic_prisoner", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go", "mon_mi_go_slaver" ],
```

Maleclypse commented on data/json/speech.json:
```diff
@@ -1495,19 +1495,19 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go", "mon_mi_go_slaver" ],
+    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_pathetic_prisoner", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go", "mon_mi_go_slaver" ],
```

Maleclypse commented on data/json/speech.json:
```diff
@@ -1495,19 +1495,19 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi_go_myrmidon" ],
+    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi_go_myrmidon", "mon_pathetic_prisoner", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi_go_myrmidon" ],
```

Maleclypse commented on data/json/speech.json:
```diff
@@ -1495,19 +1495,19 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon" ],
+    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon" ],
```

Maleclypse commented on data/json/speech.json:
```diff
@@ -1537,7 +1537,7 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon" ],
+    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go_surgeon", "mon_mi_go_guard", "mon_mi_go_myrmidon" ],
```

Maleclypse commented on data/json/speech.json:
```diff
@@ -1555,7 +1555,7 @@
   {
     "type": "speech",
-    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi-go_surgeon" ],
+    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi-go_surgeon", "mon_pathetic_prisoner", "mon_feral_prisoner" ],
```
```suggestion
    "speaker": [ "mon_mi_go", "mon_mi_go_slaver", "mon_mi-go_surgeon" ],
```

Maleclypse committed:
Apply suggestions from code review

Maleclypse commented:
This needs an interactive rebasing to eliminate the AI commits so it doesn't impact our license.  I unfortunately haven't found the time to relearn how to do that but it is on my to do list.
